pkg/controller/instance/instance_controller.go

```diff
@@ -141,12 +150,28 @@ func (r *Reconciler) Reconcile(request ctrl.Request) (ctrl.Result, error) {
 		return reconcile.Result{}, err // OV not found has to be retried because it can really have been created after Instance
 	}
 
-	// ---------- 2. First check if we should start execution of new plan ----------
+	// ---------- 2. Check if the object is being deleted, start cleanup plan ----------
+
+	var isDeleting bool
+
+	// a delete request is indicated by a non-zero 'metadata.deletionTimestamp',
+	// see https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#finalizers
+	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
```

Comment:
I don't like that part of the condition when is finalizer added (timestamp being zero) is here and part in maybeAddFinalizer (finalizer does not exist) - I would prefer to have only one place, I had to jump back and forth to make sure everything is covered.

Reply:
The main motivation is to have `maybeAddFinalizer` and `maybeRemoveFinalizer` very similar and to avoid nesting too many if-clauses.

Reply:
yeah but we still have the nesting now, just the logic is spread in two places :( I am not sure it's improving much. In that case I would rather have:

`if shouldAddFinalizer() { addFinalizer() }`
`if isBeingDeleted() { log(...) }`
and then shouldAddFinalizer() will as part of it's condition use `isBeingDeleted`

or we could just have `maybeAddFinalizer` called here and inside it would start with something like `if shouldAddFinalizer()`

I think that would make things easier for me as a reader trying to asses when we actually add finalizer :)

Reply:
That's a very good suggestion! This will make the code much easier to understand.